### PR TITLE
performance(memory): Optimize caching and refactor stack segmentation for memory efficiency

### PR DESCRIPTION
**Pull request by sedghi**
### Context

Kind of address this https://github.com/cornerstonejs/cornerstone3D/issues/492 

This PR has couple of good stuff

**1. Cache Optimizations via config enableCacheOptimization: true by default - reducing memory footprint by half going from stack to MPR and vice versa** 

Perfecting the memory consumption going to MPR from a Stack.

Basically, when we have a stack viewport (image stack) and we convert it to a volume the volume will be cached as well as the stack. However, if we can optimize this by going back to the image cache and create a view at the correct offset of the bigger volume array buffer, this will save memory. This will get enabled if cornerstone3D is configured to use SharedArrayBuffer by default 

Before
CleanShot 2024-01-02 at 10 37 49

After
CleanShot 2024-01-02 at 10 35 25

**2. Moving the decaching down, since not every volume is streaming volume but a volume should be able to decache into multiple images**

**4. Utilities for going to Stack Viewport to Volume Viewport and Vice versa**

This utility will make sure the viewports are properly handled

**5. Utilities for going from Stack Segmentation to Volume Segmentation with an example** 

#### PR

- [] My Pull Request title is descriptive, accurate and follows the
  semantic-release format and guidelines.

#### Code

- [] My code has been well-documented (function documentation, inline comments,
  etc.)

- [] I have run the `yarn build:update-api` to update the API documentation, and have
  committed the changes to this PR. (Read more here https://www.cornerstonejs.org/docs/contribute/update-api)

#### Public Documentation Updates

- [] The documentation page has been updated as necessary for any public API
  additions or removals.

#### Tested Environment

- [] "OS: 
- [] "Node version: 
- [] "Browser:
  

[blog]: https://circleci.com/blog/triggering-trusted-ci-jobs-on-untrusted-forks/
[script]: https://github.com/jklukas/git-push-fork-to-upstream-branch

### CODE REVIEW

File: common/reviews/api/core.api.md [hunk 1]
```diff
@@ -1184,6 +1224,11 @@ interface IGeometryLoadObject {
 
 // @public (undocumented)
 interface IImage {
+    // (undocumented)
+    bufferView?: {
+        buffer: ArrayBuffer;
+        offset?: number;
```
Comment (wayfarer3130) [inline common/reviews/api/core.api.md:1230]: Please require the offset if you are using a bufferView at all - that just makes things a bit more consistent, and adding in 0 is pretty easy.

File: common/reviews/api/core.api.md [hunk 2]
```diff
@@ -1212,6 +1257,8 @@ interface IImage {
     // (undocumented)
     height: number;
     // (undocumented)
+    imageFrame?: any;
```
Comment (wayfarer3130) [inline common/reviews/api/core.api.md:1260]: Please ensure this is documented as being 1 based, since it is sometimes hard to remember which one is which.

File: packages/core/src/cache/classes/ImageVolume.ts [hunk 3]
```diff
@@ -158,6 +225,453 @@ export class ImageVolume implements IImageVolume {
...
+      columnPixelSpacing: spacing[0],
+      rowPixelSpacing: spacing[1],
+      invert,
+      photometricInterpretation,
+    };
+  }
+
+  /**
+   * Converts imageIdIndex into frameIndex which will be the same
+   * for 3D volumes but different for 4D volumes
```
Comment (wayfarer3130) [inline packages/core/src/cache/classes/ImageVolume.ts:453]: Add note saying that indices are 0 based, unlike frameNumber

File: packages/core/src/eventTarget.ts [hunk 4]
```diff
@@ -12,6 +12,20 @@ class CornerstoneEventTarget implements EventTarget {
     this.listeners = {};
   }
 
+  public addEventListenerOnce(type, callback) {
```
Comment (wayfarer3130) [inline packages/core/src/eventTarget.ts:15]: Thank you - this is nice to have.

File: packages/core/src/types/Cornerstone3DConfig.ts [hunk 5]
```diff
@@ -51,6 +51,18 @@ type Cornerstone3DConfig = {
...
   };
+  /**
+   * This flag controls whether to enable cache optimization or not. Basically,
+   * when we have a stack viewport (image stack) and we convert it to a volume
+   * the volume will be cached as well as the stack. However, if we can optimize this
+   * by going back to the image cache and create a view at the correct offset
+   * of the bigger volume array buffer, this will save memory. This will get enabled
+   * if cornerstone3D is configured to use SharedArrayBuffer, the reason is that
+   * when we modify the image cache then the images are referring to a different
+   * buffer (SharedArrayBuffer) and sometimes in some system we don't want to enable
```
Comment (wayfarer3130) [inline packages/core/src/types/Cornerstone3DConfig.ts:62]: Just say that some systems don't support shared array buffer.

File: packages/core/src/types/IImage.ts [hunk 6]
```diff
@@ -120,6 +120,12 @@ interface IImage {
 
   imageQualityStatus?: ImageQualityStatus;
   calibration?: IImageCalibration;
+  imageFrame?: any;
+
+  bufferView?: {
+    buffer: ArrayBuffer;
+    offset?: number;
```
Comment (wayfarer3130) [inline packages/core/src/types/IImage.ts:127]: Make the offset required if the bufferView is defined.
Consider including the string representation of the view as well (eg 'Uint8array')

File: packages/core/src/utilities/cacheUtils.ts [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    return;
+  }
+
+  const scalarData = volume.getScalarData();
+
+  volume.imageCacheOffsetMap.size > 0
+    ? _processImageCacheOffsetMap(volume, scalarData)
+    : _processVolumeImages(volume, scalarData);
+
+  console.log(
```
Comment (wayfarer3130) [inline packages/core/src/utilities/cacheUtils.ts:54]: Remove - at least for commit.  Really we need to add a log utility so that we can leave debug information in without hitting it all the time.

File: packages/core/src/utilities/cacheUtils.ts [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    if (!image) {
+      return;
+    }
+
+    _updateImageWithScalarDataView(image, scalarData, offset);
+    cache.decrementImageCacheSize(image.sizeInBytes);
+  });
+}
+
+function _processVolumeImages(volume, scalarData) {
```
Comment (wayfarer3130) [inline packages/core/src/utilities/cacheUtils.ts:71]: Docs please

File: packages/core/src/utilities/convertVolumeToStackViewport.ts [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+  const purgeFromCache = isDerivedFromStack
+    ? allImagesAlreadyCached
+      ? true
+      : false
+    : false;
+
+  const volumeUsedInOtherViewports =
+    renderingEngine
+      .getVolumeViewports()
+      .filter((vp) => vp.hasVolumeId(volumeId)).length > 0;
```
Comment (wayfarer3130) [inline packages/core/src/utilities/convertVolumeToStackViewport.ts:83]: this can be done with find, which stops on the first element found, and does not create a nested list.

File: packages/core/src/utilities/convertVolumeToStackViewport.ts [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+  volume.decache(purgeFromCache && !volumeUsedInOtherViewports);
+
+  const stack = volume.imageIds.reverse();
```
Comment (wayfarer3130) [inline packages/core/src/utilities/convertVolumeToStackViewport.ts:87]: Use toReversed instead, so you don't mutate in place.

File: packages/core/src/utilities/convertVolumeToStackViewport.ts [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+  // their copy of the image back
+  // 2. It was actually a native volume and we need to decache it, this is a bit more
+  // complicated since then we need to decide on the imageIds for it to get
+  // decached to
+  const isDerivedFromStack = volume.imageCacheOffsetMap.size > 0;
+  const allImagesAlreadyCached = volume.imageIds.every((imageId) =>
+    cache.getImage(imageId)
+  );
+
+  const purgeFromCache = isDerivedFromStack
```
Comment (wayfarer3130) [inline packages/core/src/utilities/convertVolumeToStackViewport.ts:74]: Why not isDerivedFromStack && allImagesAlreadyCached?  I found it hard to read as a multi-level conditional.

File: packages/core/src/utilities/convertVolumeToStackViewport.ts [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    viewportId?: string;
+    background?: Types.Point3;
+  };
+}): Promise<Types.IVolumeViewport> {
+  const volumeViewport = viewport;
+  const { id, element } = volumeViewport;
+  const renderingEngine = viewport.getRenderingEngine();
+  const imageIdIndex = viewport.getCurrentImageIdIndex();
+
+  const { background } = options || { volumeId: uuidv4() };
```
Comment (wayfarer3130) [inline packages/core/src/utilities/convertVolumeToStackViewport.ts:33]: Why do you generate a volumeId here and then throw it away?  Seems very odd.  You aren't even allowed to specify  volumeId in the options type.

File: packages/core/src/utilities/convertVolumeToStackViewport.ts [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+ * @param params - The parameters for the conversion.
+ * @param params.viewport - The volume viewport to convert.
+ * @param params.options - The conversion options.
+ * @param [params.options.viewportId] - The new stackViewportId, If not provided, the volume viewport id will be used.
+ * @param [params.options.background] - The background color of the stack viewport.
+ * @param [params.options.decache] - Whether to decache the volume. Defaults to false.
+ *
+ * @returns The converted stack viewport.
+ */
+async function convertVolumeToStackViewport({
```
Comment (wayfarer3130) [inline packages/core/src/utilities/convertVolumeToStackViewport.ts:18]: Should this throw if the viewport is showing a non-acquisition plane modality?

File: packages/core/src/utilities/convertVolumeToStackViewport.ts [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+  volume.decache(purgeFromCache && !volumeUsedInOtherViewports);
+
+  const stack = volume.imageIds.reverse();
+  await stackViewport.setStack(
+    stack,
+    Math.max(volume.imageIds.length - imageIdIndex - 1, 0)
+  );
+
+  // Render the image
+  volumeViewport.render();
```
Comment (wayfarer3130) [inline packages/core/src/utilities/convertVolumeToStackViewport.ts:94]: Why are you rendering the volume viewport?  Won't that be the wrong one now?

File: packages/streaming-image-volume-loader/src/BaseStreamingImageVolume.ts [hunk 15]
```diff
@@ -351,6 +241,17 @@ export default class BaseStreamingImageVolume
     // it is either cachedImage or cachedVolume
     const isFromImageCache = !!cachedImage;
 
+    if (isFromImageCache) {
```
Comment (wayfarer3130) [inline packages/streaming-image-volume-loader/src/BaseStreamingImageVolume.ts:244]: Can you check the target buffer too?  If that isn't set, you shouldn't be doing anything here.

File: common/reviews/api/core.api.md (hunk 2 above)
Comment (sedghi) [reply common/reviews/api/core.api.md:1260]: it is the imageFrame object from the CSWIL

File: packages/core/src/utilities/convertVolumeToStackViewport.ts (hunk 10 above)
Comment (sedghi) [reply packages/core/src/utilities/convertVolumeToStackViewport.ts:87]: I see it is pretty recent

CleanShot 2024-01-02 at 12 12 49

I will write it manually

File: packages/tools/examples/stackToVolumeWithSegmentations/index.ts [hunk 16]
```diff
@@ -0,0 +1,430 @@
...
+  },
+});
+
+function addOrientationDropdownIfVolumeViewport() {
+  const renderingEngine = getRenderingEngine(renderingEngineId);
+  const viewport = renderingEngine.getViewport(viewportId);
+
+  // Check if the current viewport is a VolumeViewport
+  if (viewport.type === ViewportType.ORTHOGRAPHIC) {
+    addDropdownToToolbar({
```
Comment (wayfarer3130) [inline packages/tools/examples/stackToVolumeWithSegmentations/index.ts:167]: There is no corresponding remove when you drop it back to stack, so you get more and more dropdowns as you add.

File: packages/core/src/utilities/convertVolumeToStackViewport.ts (hunk 11 above)
Comment (sedghi) [reply packages/core/src/utilities/convertVolumeToStackViewport.ts:74]: This is a terrible API, unfortunately, that I will fix in 2.x. Basically, `decache(true)` means remove from cache, `decache(false)` means decache into images. It doesn't make sense, but it is what it is for now.

File: packages/core/src/utilities/convertVolumeToStackViewport.ts (hunk 13 above)
Comment (sedghi) [reply packages/core/src/utilities/convertVolumeToStackViewport.ts:18]: Hmmm, I don't think so, since it is only a `view` and does not matter regarding the underlying images

**Comment by wayfarer3130** [comment]
Can get into a very weird state with the display going orange if you:
yarn run example stacktovolumewithsegmentations
Disable cache/set Fast 5g (50mbit/s, 10 ms latency)
Convert/add segmentation randomly between stack and volume on different slices.
Wait for load to complete
Sometimes, but not always, nothing is being loaded, but images are all gray/blank

Suggest configuring this with the interleaved loading and test with that as well - this should still work, and eventually succeed.

File: packages/tools/src/eventListeners/segmentation/imageChangeEventListener.ts [hunk 17]
```diff
@@ -148,10 +150,25 @@ function _imageChangeEventListener(evt) {
...
+    const { origin: currentOrigin } =
+      viewport.getImageDataMetadata(currentImage);
+
+    // IMPORTANT: We need to make sure that the origin of the segmentation
+    // is the same as the current image origin. This is because due to some
+    // floating point precision issues, when coming from volume to stack
+    // the origin of the segmentation can be slightly different from the
+    // current image origin. This can cause the segmentation to be rendered
+    // in the wrong location.
+    // Todo: This will not work for segmentations that are not in the same
```
Comment (wayfarer3130) [inline packages/tools/src/eventListeners/segmentation/imageChangeEventListener.ts:166]: I think you mean in the same Frame of Reference?

File: packages/tools/examples/stackToVolumeWithSegmentations/index.ts (hunk 16 above)
Comment (sedghi) [reply packages/tools/examples/stackToVolumeWithSegmentations/index.ts:167]: There is down below

```js
const orientationDropdown = document.getElementById('orientationDropdown');
    if (orientationDropdown) {
      orientationDropdown.remove();
    }
```

**Comment by wayfarer3130** [comment]
> 3. Refactor the Stack segmentation to accept multiple stack of images
>
> This is building block of enabling rendering of the overlapping segments. The rendering part is not there yet, but will be soon.

I'm not convinced that is a good idea - it makes knowing which segmentation instance to draw into quite difficult.  There are several other options where a segment index can represent multiple segments that seems like a better solution.  The memory consumption of that is also quite a bit less.  Having multiple segmentation images should be handled by having two or more segmentation UIDs, each with a single segmentation image per base image frame.

File: packages/tools/src/eventListeners/segmentation/segmentationDataModifiedEventListener.ts [hunk 18]
```diff
@@ -101,7 +101,10 @@ const onSegmentationDataModified = function (
           const currentSegmentationImageId =
             imageIdReferenceMap.get(currentImageId);
 
-          const segmentationImage = cache.getImage(currentSegmentationImageId);
+          // Todo: fix the logic to support multiple derived imageIds
+          const segImageId = [...currentSegmentationImageId][0];
```
Comment (wayfarer3130) [inline packages/tools/src/eventListeners/segmentation/segmentationDataModifiedEventListener.ts:105]: This seems like a bad idea to convert to an array and back - it is relatively slow - you can instead get the iterator if you want and the first element, or add a helper function to do that for you.

File: packages/tools/src/store/ToolGroupManager/ToolGroup.ts [hunk 19]
```diff
@@ -185,6 +185,10 @@ export default class ToolGroup implements IToolGroup {
    * @param renderingEngineId - The rendering engine to use.
    */
   public addViewport(viewportId: string, renderingEngineId?: string): void {
+    if (typeof viewportId !== 'string') {
```
Comment (wayfarer3130) [inline packages/tools/src/store/ToolGroupManager/ToolGroup.ts:188]: How about !viewportId  since that checks for empty, null and undefined, which seems a better solution here.

Comment (sedghi) [reply packages/tools/src/store/ToolGroupManager/ToolGroup.ts:188]: I was passing the `viewport` object itself, and I wasted 30 minutes on this trying to figure out what was wrong. So, `!viewportId` will not catch that, unfortunately.

File: packages/tools/src/tools/displayTools/Labelmap/addLabelmapToElement.ts [hunk 20]
```diff
@@ -61,10 +61,13 @@ async function addLabelmapToElement(
   } else {
     // We can use the current imageId in the viewport to get the segmentation imageId
     // which later is used to create the actor and mapper.
-    const segmentationImageId = (
+    const segmentationImageIds = (
       labelMapData as LabelmapSegmentationDataStack
     ).imageIdReferenceMap.get(viewport.getCurrentImageId());
 
+    // Todo: fix this to support multiple derived imageIds
```
Comment (wayfarer3130) [inline packages/tools/src/tools/displayTools/Labelmap/addLabelmapToElement.ts:68]: Can you revert the change and instead allow multiple segmentation UIDs to be associated?  That way they are cleanly separated.  With this change it becomes very hard to know where to do things.

File: packages/tools/src/tools/segmentation/BrushTool.ts [hunk 21]
```diff
@@ -69,7 +69,7 @@ class BrushTool extends BaseTool {
   private _editData: {
     segmentsLocked: number[]; //
     segmentationRepresentationUID?: string;
-    imageIdReferenceMap?: Map<string, string>;
+    imageIdReferenceMap?: Map<string, Set<string>>;
```
Comment (wayfarer3130) [inline packages/tools/src/tools/segmentation/BrushTool.ts:72]: It really isn't obvious to me how you would figure out which one is of interest.  I think they should be keyed by the segmentation UID, which would mean you get exactly one value, so that you would then undo this line.

File: packages/tools/src/tools/segmentation/BrushTool.ts [hunk 22]
```diff
@@ -235,11 +233,9 @@ class BrushTool extends BaseTool {
       // and should throw an error or maybe simply just allow circle manipulation
       // and not sphere manipulation
       if (this.configuration.activeStrategy.includes('SPHERE')) {
```
Comment (wayfarer3130) [inline packages/tools/src/tools/segmentation/BrushTool.ts:235]: Can you assign an error message on the activeStrategy object that can be thrown instead?  Eg, something like:
if( this.getStrategy().VOLUMETRIC_ERROR_MESSAGE ) {
throw new Error(this.getStrategy().VOLUMETRIC_ERROR_MESSAGE );

That way you don't need to know anything special about the strategy, just that it doesn't work unless volumetric.  I really hate including the string key here as it makes it very brittle on changes or additions.

File: packages/tools/src/tools/segmentation/strategies/fillCircle.ts [hunk 23]
```diff
@@ -98,9 +95,9 @@ function createPointInEllipse(worldInfo: {
 
   const radius = Math.max(xRadius, yRadius, zRadius);
   if (
-    nearOrZero(xRadius, radius) &&
-    nearOrZero(yRadius, radius) &&
-    nearOrZero(zRadius, radius)
+    csUtils.isEqual(xRadius, radius) &&
```
Comment (wayfarer3130) [inline packages/tools/src/tools/segmentation/strategies/fillCircle.ts:98]: Extract isEqual up on line 19 or so - this function is called enough times that you don't want to dereference the csUtils multiple times.

File: packages/tools/src/tools/segmentation/strategies/utils/getStrategyData.ts [hunk 24]
```diff
@@ -40,7 +40,10 @@ function getStrategyData({ operationData, viewport }) {
     // of that actor at that moment so we have the imageData already
     const actor = viewport.getActor(segmentationRepresentationUID);
     segmentationImageData = actor.actor.getMapper().getInputData();
-    const currentSegmentationImageId = imageIdReferenceMap.get(currentImageId);
+    const currentSegmentationImageIds = imageIdReferenceMap.get(currentImageId);
+
+    // Todo: fix this to support multiple derived imageIds
+    const currentSegmentationImageId = [...currentSegmentationImageIds][0];
```
Comment (wayfarer3130) [inline packages/tools/src/tools/segmentation/strategies/utils/getStrategyData.ts:46]: Leaving this code everywhere without having real implementations seems like a bad idea for putting this in core.  I'd rather see the old code here and then actually fix it once we understand what is actually going to happen.

File: packages/tools/src/types/LabelmapTypes.ts [hunk 25]
```diff
@@ -42,13 +42,10 @@ export type LabelmapSegmentationDataVolume = {
...
-   * This is a Map from referenced imageId to the segmentation (Derived) imageId that
-   * has been used to create the derived labelmap.
-   * Todo: later I guess we can have it as Record<string, metadata> where metadata
-   * can contain a derived image arbitrary information, for use cases such that the labelmap is
-   * derived from another image that is irrelevant to the current viewport.
+   * This is a Map from referenced imageId to the segmentation (Derived) imageIds (can be
+   * multiple) that are associated with it.
    */
-  imageIdReferenceMap: Map<string, string>;
+  imageIdReferenceMap: Map<string, Set<string>>;
```
Comment (wayfarer3130) [inline packages/tools/src/types/LabelmapTypes.ts:48]: Ditto - remove this

**Review by wayfarer3130 — APPROVED** [review]
I'm ok with the changes as they are now - a couple of things could still be improved, but I think it is good enough to go as is.
